pravega/pravega
# Issue 284: Controller's workflow for stream truncation

**Change log description**
Implements controller side workflow for processing a stream truncation request. 

Controller has a new metadata called StreamTruncationRecord which corresponds to latest stream cut that was processed. 

When a call comes to controller to perform truncation workflow, it first computes the StreamTruncationRecord, this data structure has information to help compute available segments for getActiveSegments calls. 

Following is the flow of information:
1. controllerService.truncateStream(streamCut)
2. streamMetadataTasks.truncateStream(streamCut)
-compute new StreamTruncationRecord corresponding to streamCut. identify segments to delete 
-post truncate event
-try update store.truncationRecord with updating = true.
3. pick event, if truncationRecord.updating, process truncation:
- call segment store to delete purged segments
- call segment store to truncate segments in stream cut
- complete truncationRecord updating = false, clear `toDelete` segments and merge them in deleted segments field. 

A new abstraction of StreamProperty is added over stream properties that can be updated via a workflow. Namely `configuration` and `truncationRecord`. 
StreamProperty is a wrapper over actual property with an `updating` flag that indicates when a property is bring updated. 
Same logic has been adopted for updateConfiguration where configuraiton update happens inline in caller's update request's context. Update Event processing looks for `updating` flag in   

So the event for truncation and update stream no longer need to contain the update to apply. 
The update is applied by the caller, and if they are able to apply the update successfully, there will be an event posted in requeststream for processing the update. 
This event's processing will process if the property is `updating` and perform rest of the workflow required to complete the update. 
This has twofold advandates for update: 
1. removes the logic of ConfigWithVersion as properties no longer need to maintain versions inside the metadata.
2. Fixes races where two callers can concurrently invoke the same workflow and all scenarios are possible - both succeed(one applied after another), first fails or seconds fails. 
That was causing a unit test to fail sporadically.
With this change, its more predictable which update will be processed as with this approach there can only be exactly one pending update.   

**Purpose of the change**
Subtask of #284.

**What the code does**
Implements truncation workflow. 

**How to verify it**
Unit tests added
